### PR TITLE
Plane: added Q_LAND_ALTCHG parameter

### PR DESCRIPTION
**Pull request by tridge**
this is the threshold height change over 4 seconds for a landing to be
detected. It can be raised if landing detection is very slow

### CODE REVIEW

**Review by peterbarker — APPROVED** [review]

**Comment by peterbarker** [comment]
`Q_LAND_ALTCHG_MAX` maybe?

File: ArduPlane/quadplane.h [hunk 1]
```diff
@@ -329,6 +329,9 @@ class QuadPlane
     AP_Int8 enable;
     AP_Int8 transition_pitch_max;
 
+    // landing detection threshold in meters
+    AP_Float land_detect_threshold;
```
Comment (IamPete1) [inline ArduPlane/quadplane.h:333]: Maybe put it in the struct

https://github.com/ArduPilot/ardupilot/blob/54208574f0ee96b406a854069d787d34cfa6d5f7/ArduPlane/quadplane.h#L400-L405

File: ArduPlane/quadplane.cpp [hunk 2]
```diff
@@ -2995,7 +3004,7 @@ bool QuadPlane::land_detector(uint32_t timeout_ms)
     // we only consider the vehicle landed when the motors have been
     // at minimum for timeout_ms+1000 and the vertical position estimate has not
     // changed by more than 20cm for timeout_ms
-    if (fabsf(height - landing_detect.vpos_start_m) > 0.2) {
+    if (fabsf(height - landing_detect.vpos_start_m) > land_detect_threshold) {
```
Comment (IamPete1) [inline ArduPlane/quadplane.cpp:3007]: This resets `land_start_ms` which is then compared against the timeout passed in. That is hard codded to 4 seconds, I wonder if we would be better to make that timeout a parameter rather than this distance. The timeout is use for both the lower motor limit check and this movement check. So that timeout would give more control.

File: ArduPlane/quadplane.cpp [hunk 3]
```diff
@@ -434,6 +434,15 @@ const AP_Param::GroupInfo QuadPlane::var_info2[] = {
...
     AP_SUBGROUPPTR(weathervane, "WVANE_", 30, QuadPlane, AC_WeatherVane),
 
+    // @Param: LAND_DET_THRS
+    // @DisplayName: Land detection threshold
+    // @Description: The maximum altitude change allowed during land detection. You can raise this value if you find that landing detection takes a long time to complete. It is the maximum change in altitude over a period of 4 seconds for landing to be detected
+    // @Units: m
+    // @Range: 0.1 0.6
+    // @Increment: 0.05
+    // @User: Standard
+    AP_GROUPINFO("LAND_DET_THRS", 31, QuadPlane, land_detect_threshold, 0.2),
```
[the remark this answers is not in the document]
Comment (IamPete1) [reply ArduPlane/quadplane.cpp:444]: I agree.

**Comment by IamPete1** [comment]
I do think that the issue this fixes is still fundamentally a EKF one, we have not had any other reports as far as I know, this landing detection method used to work fine.

**Comment by pompecukor** [comment]
> I do think that the issue this fixes is still fundamentally a EKF one, we have not had any other reports as far as I know, this landing detection method used to work fine.

I agree. Everything was fine, then there was some change as I said somewhere between September and December last year. something in the master that brought this about. All out craft disarmed fine without any issue before that. We are talking not even just one type of platform, but both out Vtol systems, and they are very different.
What I too find strange is why there is not other such report. However I am not sure how many people fly latest master from about December and do full auto mission. I would think that most testers and DIY guys that I know who take the risk to run cutting edge are usually landing and disarming manually, not mission.
This PR comes to mind as a perfect example. No one noticed/reported it except us, meanwhile it was actually affecting almost anyone using the same firmware. https://github.com/ArduPilot/ardupilot/pull/19626 And the issue was in there (in the firmware) for a few months, that was when I started to question how many people were actually ineptly running and testing master VTOL all round. I think most test are targeted to test and tune new features, hence not comprehensive enough.

File: ArduPlane/quadplane.h (hunk 1 above)
Comment (tridge) [reply ArduPlane/quadplane.h:333]: done

File: ArduPlane/quadplane.cpp (hunk 2 above)
Comment (tridge) [reply ArduPlane/quadplane.cpp]: i prefer to leave the motor limit checks as 4s. I think adjusting height change threshold is the easiest way to handle the issue, outside of EKF changes

File: ArduPlane/quadplane.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (tridge) [reply ArduPlane/quadplane.cpp]: changed

**Comment by pompecukor** [comment]
> Alternatively if a replayable log can be provided it is likely that we can make some tuning changes to more closely track the Baro such as reduce the default value of EK3_ALT_M_NSE.
> 

I have shared logs with replay with you and devs in discord.

**Comment by tridge** [comment]
@priseborough a log is here:
http://uav.tridgell.net/tmp/2022-03-07_Ship_landing_firmware.bin
it is a replay log, but does not have LOG_DISARMED. It only has EK3 enabled, but I don't think this issue is EKF3 only
It also does not impact all lanes equally in this log:
image
I still think the change in this PR is worthwhile to give users a way to adjust this behaviour, even if we find a tuning change for the EKF

**Comment by pompecukor** [comment]
@tridge 

> @priseborough a log is here: http://uav.tridgell.net/tmp/2022-03-07_Ship_landing_firmware.bin it is a replay log, but does not have LOG_DISARMED. It only has EK3 enabled, but I don't think this issue is EKF3 only It also does not impact all lanes equally in this log: image I still think the change in this PR is worthwhile to give users a way to adjust this behaviour, even if we find a tuning change for the EKF

I already sent log to him yesterday. Anyway I was not aware Log_disarm was needed, it was not in his instructions. Also I imagined that this issue was as related to after arming, so did not know it would matter.
About the EKF2, he said "alternately to do Log_replay" so I imagined it was one or the other.
If you like and it helps, I am happy to enable, but not use EKF2 also and set Log_disarm to yes. 
Then do a couple more landings.
Let me know if you need any of that.

**Comment by priseborough** [comment]
Here are the vertical velocity, vertical position, baro alt and EKF alt estimate graphs from EKF2 and EKF3 for one of the affected landings. Also shown is the baro pressure alt and GPS alt. This shows that the underlying issue is that when the lift motors ar running the

EKF2 innovations and baro height tracking show worse tracking to baro during descent than for EKF3.

Screen Shot 2022-03-09 at 7 10 55 am

EKF3 innovations and baro height tracking shows that use of a smaller EK3_GND_EFF_DZ   = 1.0  (down from 4.0) has given better tracking of baro height during decent, but the EKF is correcting for a baro height difference after the motors reduce which results in the height estimate being pulled down to match the baro.

Comparing GPS, baro and EKF heights shows the underlying issue is that the baro has a different offset in air than on ground so the EKF has to correct after touchdown. 

Screen Shot 2022-03-09 at 7 41 02 am

The log was then replayed with the default value of EK3_GND_EFF_DZ and although the initial height drift rate after touchdown is less, eventually the EKF still has to match the Baro.

Screen Shot 2022-03-09 at 7 41 02 am

In conclusion this is not an EKF algorithm or tuning issue. The landing detector thresholds and or logic need changing.

**Comment by tridge** [comment]
> In conclusion this is not an EKF algorithm or tuning issue. The landing detector thresholds and or logic need changing.

thanks for the analysis Paul! I'll merge this PR which gives one simple parameter to change to reduce the issue
@pompecukor please try setting Q_LAND_ALTCHG to 0.4 and you should fine the issue is greatly reduced
